### PR TITLE
[TTAHUB-2238] Filter for recipients without activity in range

### PR DESCRIPTION
**Pull request by AdamAdHocTeam**
## Description of change

This adds a new filter to the RTR search called 'Recipients without TTA'. 

Activity is determined by using the start and end date of the linked AR's.

image

It has the following options: 
- Last thirty days
- Last three months
- Last six months
- Year to date
- Custom date range which allows users to enter a start and end date. 

**Note:** There is a linked mockup that shows a completely different layout of the date filter. If we need this change also it should be its own ticket as its not in this JIRA's requirements.

## How to test

- Review the code
- All tests pass
- Test the new filter from the RTR search
- Make sure the existing date filters worked as they did before.

Something like the below query can be used to find activity in the range (the opposite).
`  SELECT
	  	ar.id,
		ar."startDate",
		ar."endDate"
	  FROM "ActivityReports" ar WHERE id IN (
	  SELECT "activityReportId" 
	  FROM "ActivityReportGoals"
	  WHERE "goalId" IN (
	  SELECT id from "Goals" WHERE "grantId" = (
		 SELECT id FROM "Grants" WHERE "recipientId" IN (
		  SELECT id FROM "Recipients" WHERE "name" = 'City of New Haven, Inc.'
	  ) AND status = 'Active')))
	  	 AND (ar."startDate", ar."endDate") OVERLAPS ('01/01/2022','11/28/2023')
	  order by ar."startDate", ar."endDate"`

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-2338

### Every PR

- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [x] Update JIRA ticket status

### CODE REVIEW

File: src/scopes/grants/recipientsWithoutTTA.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+/* eslint-disable import/prefer-default-export */
+import { Op } from 'sequelize';
+import { sequelize } from '../../models';
+
+const grantsMissingActivitySql = (beginActivityDate, finishActivityDate) => sequelize.literal(
+  `
+    -- Determine what grants have activity reports within the given date range.
```
Comment (thewatermethod) [inline src/scopes/grants/recipientsWithoutTTA.js:7]: can we make these JS comments? That way they don't get printed in the server logs everytime someone uses the filters

**Comment by thewatermethod** [comment]
Is this right? I seached for recipients without TTA this YTD.

Screenshot 2023-11-28 at 12 51 43 PM

I looked at the TTA history for this first recipient and they have 28 reports this year.

Screenshot 2023-11-28 at 12 52 12 PM

**Comment by AdamAdHocTeam** [comment]
> grantsMissingActivitySql

Yea they want everything in one section as opposed to what we have now:

image

**Comment by AdamAdHocTeam** [comment]
> Is this right? I seached for recipients without TTA this YTD.
> 
> Screenshot 2023-11-28 at 12 51 43 PM
> 
> I looked at the TTA history for this first recipient and they have 28 reports this year.
> 
> Screenshot 2023-11-28 at 12 52 12 PM

I think this is correct as it would be all grants that meet two pieces of criteria:
1. Status active
2.    OR (
          G."endDate" > '2020-08-31 04:00:00.000 +00:00' 
          AND (
            G."inactivationDate" IS NULL 
            OR G."inactivationDate" > '2020-08-31 04:00:00.000 +00:00'
          )
        )
A rough query shows about 800 before the other filters are applied.

**Comment by thewatermethod** [comment]
> I think this is correct as it would be all grants that meet two pieces of criteria:

Confused by this. If a recipient had TTA this year I would think they would not show up in the filtered data. How do the grant scopes relate?

File: src/scopes/grants/recipientsWithoutTTA.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+*/
+const grantsMissingActivitySql = (beginActivityDate, finishActivityDate) => sequelize.literal(
+  `
+    (WITH activity AS (
+      SELECT
+        g."id" AS used_grant_id
+      FROM "Grants" g
+      JOIN "Goals" gl
+        ON g.id = gl."grantId"
+      JOIN "ActivityReportGoals" arg
```
Comment (thewatermethod) [inline src/scopes/grants/recipientsWithoutTTA.js:18]: shouldn't this join with ActivityRecipients, why are we using goals?

Comment (thewatermethod) [reply src/scopes/grants/recipientsWithoutTTA.js:18]: I guess maybe it doesn't matter

**Comment by AdamAdHocTeam** [comment]
> Is this right? I seached for recipients without TTA this YTD.
> 
> Screenshot 2023-11-28 at 12 51 43 PM
> 
> I looked at the TTA history for this first recipient and they have 28 reports this year.
> 
> Screenshot 2023-11-28 at 12 52 12 PM

I cant repro this can you make sure you have latest as there have been some small (critiical) changes.

File: src/scopes/grants/recipientsWithoutTTA.js (hunk 2 above)
Comment (AdamAdHocTeam) [reply src/scopes/grants/recipientsWithoutTTA.js:18]: Good call

**Comment by thewatermethod** [comment]
> I cant repro this can you make sure you have latest as there have been some small (critical) changes.

Restarting my app, refreshed DB. 

I am still seeing the same thing. To repro:
1) go to the rtr search
2) select Recipients without TTA is year to date
3) click first recipient in list
4) go to tta history tab
5) filter should be "date started is year to date" but if not, select that. TTA history filters ARs.
6) they have 20-odd reports. Scrolling down, the table has reports with start dates and end dates within this year.

**Review by thewatermethod — APPROVED** [review]
Looks good

**Pull request merged**